File: python/euler/bin/p0093.py
# project euler: problem 93

# combination of numberes:
#   nCk = C(n,k) = C(9,4) = 9*8*7*6 / 4*3*2*1 = 126
#
# arithmetic operations (four_ops):
#   commutative:
#     addition: X + Y
#     multiplication: X * Y
#   no-commutative:
#     subtraction:  X - Y, Y - X
#     division: X / Y, Y / X
#
# patterns:
#   A, B, C, D: numbers
#
#   [1] ((A op B) op C) op D  [C(4,2)*2 = 12]
#        ^^^--^^^
#       ^^^^^^^^^^--^^^
#       ^^^^^^^^^^^^^^^^--^^
#   [2] (A op B) op (C op D)  [C(4,2) = 6]
#       ^^^--^^^    ^^^--^^^
#       ^^^^^^^^^--^^^^^^^^^
#
#   ^-^: We can ignore the order of the two terms because
#        four_ops() considers no-commutative operations.
#
# total(?):
#     C(9,4) * 6^3 [three arithmetic operattions] * (C(4,2)*2 + C(4,2))
#   = 126 * 216 * 18
#   = 489888

from fractions import Fraction
from itertools import combinations, count, product


def four_ops(x1: Fraction, x2: Fraction) -> list[Fraction]:
    result = [x1 + x2, x1 * x2, x1 - x2, x2 - x1]
    if x1 != 0:
        result.append(x2 / x1)
    if x2 != 0:
        result.append(x1 / x2)

    return result


def case_1(a: Fraction, b: Fraction, lst: list[Fraction]) -> list[Fraction]:
    # ((A op B) op C) op D
    result = []
    for ab in four_ops(a, b):
        # C: lst[0], D: lst[1]
        for abc in four_ops(ab, lst[0]):
            result += four_ops(abc, lst[1])
        # C: lst[1], D: lst[0]
        for abc in four_ops(ab, lst[1]):
            result += four_ops(abc, lst[0])

    return result


def case_2(a: Fraction, b: Fraction, lst: list[Fraction]) -> list[Fraction]:
    # (A op B) op (C op D)
    result = []
    ab = four_ops(a, b)
    cd = four_ops(lst[0], lst[1])

    for tpl in product(ab, cd):
        result += four_ops(tpl[0], tpl[1])

    return result


def make_numbers(tpl: tuple[Fraction, Fraction, Fraction, Fraction]) -> set[int]:
    lst = []
    for i in range(len(tpl)):
        for j in range(i + 1, len(tpl)):
            rest = []
            for k in range(len(tpl)):
                if k not in (i, j):
                    rest.append(tpl[k])

            lst += case_1(tpl[i], tpl[j], rest)
            lst += case_2(tpl[i], tpl[j], rest)

    return set(x.numerator for x in lst if x.denominator == 1)


def get_consecutive_length(tpl: tuple[Fraction, Fraction, Fraction, Fraction]) -> int:
    nums = make_numbers(tpl)
    for cnt in count(1):
        if cnt not in nums:
            break

    return cnt - 1


def compute() -> str:
    max_length = 0
    for tpl in combinations([Fraction(i) for i in range(1, 10)], 4):
        if (tmp := get_consecutive_length(tpl)) > max_length:
            answer = tpl
            max_length = tmp

    return ''.join(str(frac.numerator) for frac in answer)


def solve() -> str:
    return compute()

File: typescript/euler/bin/p0093.ts
// project euler: problem 93

/*
combination of numberes:
  nCk = C(n,k) = C(9,4) = 9*8*7*6 / 4*3*2*1 = 126

arithmetic operations (four_ops):
  commutative:
    addition: X + Y
    multiplication: X * Y
  no-commutative:
    subtraction:  X - Y, Y - X
    division: X / Y, Y / X

patterns:
  d1, d2, d3, d4: numbers

  [case 1]
     ((d1 OP d2) OP d3) OP d4
      ^^^----^^^
     ^^^^^^^^^^^----^^^
     ^^^^^^^^^^^^^^^^^^----^^

     ((d1 OP d2) OP d4) OP d3
      ^^^----^^^
     ^^^^^^^^^^^----^^^
     ^^^^^^^^^^^^^^^^^^----^^
  [case 2]
     (d1 OP d2) OP (d3 OP d4)
     ^^^----^^^    ^^^----^^^
     ^^^^^^^^^^----^^^^^^^^^^

  ^^-^^: We can ignore the order of the two terms because
         four_ops() considers no-commutative operations.
*/

import { cartesianProduct, combinations } from "combinatorics/mod.ts";
import { Ratio, rational } from "../lib/rational.ts";

const four_ops = (x1: Ratio, x2: Ratio): Ratio[] => {
  if (x1.isZero()) {
    return [x2.negate(), rational(0), x2];
  }

  if (x2.isZero()) {
    return [x1.negate(), rational(0), x1];
  }

  return [
    x1.add(x2),
    x1.mul(x2),
    x1.sub(x2),
    x2.sub(x1),
    x1.div(x2),
    x2.div(x1),
  ];
};

const case_1 = (x1: Ratio, x2: Ratio, rest: Ratio[]): Ratio[] => {
  let result: Ratio[] = [];
  for (const d1d2 of four_ops(x1, x2)) {
    // d3: rest[0], d4: rest[1]
    for (const d1d2d3 of four_ops(d1d2, rest[0])) {
      result = result.concat(four_ops(d1d2d3, rest[1]));
    }
    for (const d1d2d4 of four_ops(d1d2, rest[1])) {
      result = result.concat(four_ops(d1d2d4, rest[0]));
    }
  }

  return result;
};

const case_2 = (x1: Ratio, x2: Ratio, rest: Ratio[]): Ratio[] => {
  let result: Ratio[] = [];
  for (
    const [d1d2, d3d4] of cartesianProduct(
      four_ops(x1, x2),
      four_ops(rest[0], rest[1]),
    )
  ) {
    result = result.concat(four_ops(d1d2, d3d4));
  }

  return result;
};

const make_numbers = (lst: Ratio[]): Set<number> => {
  // The parameter `lst` must be a Ratio list of four elements.
  const result = new Set<number>();
  for (let i = 0; i < lst.length - 1; i++) {
    for (let j = i + 1; j < lst.length; j++) {
      const rest: Ratio[] = [];
      for (let k = 0; k < lst.length; k++) {
        if (k !== i && k !== j) {
          rest.push(lst[k]);
        }
      }

      // [1] ((d1 OP d2) OP d3) OP d4
      //     ((d1 OP d2) OP d4) OP d3
      for (const rat of case_1(lst[i], lst[j], rest)) {
        if (rat.isInteger()) {
          result.add(Number(rat.num));
        }
      }

      // [2] (d1 OP d2) OP (d3 OP d4)
      if (i !== 0) {
        continue;
      }
      for (const rat of case_2(lst[i], lst[j], rest)) {
        if (rat.isInteger()) {
          result.add(Number(rat.num));
        }
      }
    }
  }

  return result;
};

const count_consec_numbers = (lst: Ratio[]): number => {
  const n_set = make_numbers(lst);
  let cnt = 1;
  while (n_set.has(cnt)) {
    cnt += 1;
  }

  return cnt - 1;
};

export const compute = (): string => {
  let max_count = 0;
  let nums: Ratio[] = [];
  const rational_numbers = [
    rational(1),
    rational(2),
    rational(3),
    rational(4),
    rational(5),
    rational(6),
    rational(7),
    rational(8),
    rational(9),
  ];

  for (const lst of combinations(rational_numbers, 4)) {
    const len = count_consec_numbers(lst);
    if (len > max_count) {
      max_count = len;
      nums = lst;
    }
  }

  let result = 0;
  for (const x of nums) {
    result = result * 10 + Number(x.num);
  }

  return String(result);
};

export const solve = (): string => compute();
